File: src/main/java/me/lokka30/levelledmobs/commands/subcommands/KillSubcommand.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.commands.subcommands;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.commands.MessagesBase;
import me.lokka30.levelledmobs.misc.RequestedLevel;
import me.lokka30.levelledmobs.util.Utils;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.command.BlockCommandSender;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.entity.Tameable;
import org.bukkit.entity.ZombieVillager;
import org.bukkit.metadata.FixedMetadataValue;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Allows you to kill LevelledMobs with various options including all levelled mobs, specific worlds
 * or levelled mobs in your proximity
 *
 * @author stumper66
 * @since 2.0
 */
public class KillSubcommand extends MessagesBase implements Subcommand {

    public KillSubcommand(final LevelledMobs main) {
        super(main);
    }

    @Override
    public void parseSubcommand(final LevelledMobs main, final @NotNull CommandSender sender,
        final String label, final String[] args) {
        commandSender = sender;
        messageLabel = label;

        if (!sender.hasPermission("levelledmobs.command.kill")) {
            main.configUtils.sendNoPermissionMsg(sender);
            return;
        }

        if (args.length <= 1) {
            showMessage("command.levelledmobs.kill.usage");
            return;
        }

        int checkArgs = args.length;
        boolean useNoDrops = false;
        final RequestedLevel rl = getLevelFromCommand(sender, args);
        if (rl != null) {
            if (rl.hadInvalidArguments) {
                return;
            }
            checkArgs -= 2;
        }

        for (final String arg : args) {
            if ("/nodrops".equalsIgnoreCase(arg)) {
                useNoDrops = true;
                break;
            }
        }

        switch (args[1].toLowerCase()) {
            case "all":
                if (!sender.hasPermission("levelledmobs.command.kill.all")) {
                    main.configUtils.sendNoPermissionMsg(sender);
                    return;
                }

                if (checkArgs == 2) {
                    if (sender instanceof Player) {
                        final Player player = (Player) sender;
                        parseKillAll(List.of(player.getWorld()), main, useNoDrops, rl);
                    } else {
                        showMessage("command.levelledmobs.kill.all.usage-console");
                    }

                } else if (checkArgs == 3 || checkArgs == 4) {
                    if (args[2].equals("*")) {
                        parseKillAll(Bukkit.getWorlds(), main, useNoDrops, rl);
                        return;
                    }

                    if ("/nodrops".equalsIgnoreCase(args[2])) {
                        parseKillAll(Bukkit.getWorlds(), main, true, rl);
                    } else {
                        final World world = Bukkit.getWorld(args[2]);
                        if (world == null) {
                            showMessage("command.levelledmobs.kill.all.invalid-world", "%world%",
                                args[2]);
                            return;
                        }
                        parseKillAll(List.of(world), main, useNoDrops, rl);
                    }
                } else {
                    showMessage("command.levelledmobs.kill.all.usage");
                }
                break;
            case "near":
                if (!sender.hasPermission("levelledmobs.command.kill.near")) {
                    main.configUtils.sendNoPermissionMsg(sender);
                    return;
                }

                if (checkArgs == 3 || checkArgs == 4) {
                    if (!(sender instanceof BlockCommandSender) && !(sender instanceof Player)) {
                        showMessage("common.players-only");
                        return;
                    }

                    int radius;
                    try {
                        radius = Integer.parseInt(args[2]);
                    } catch (final NumberFormatException exception) {
                        showMessage("command.levelledmobs.kill.near.invalid-radius", "%radius%",
                            args[2]);
                        //messages = Utils.replaceAllInList(messages, "%radius%", args[2]); //After the list is colourised, so %radius% is not coloursied.
                        return;
                    }

                    final int maxRadius = 1000;
                    if (radius > maxRadius) {
                        radius = maxRadius;
                        showMessage("command.levelledmobs.kill.near.invalid-radius-max",
                            "%maxRadius%", String.valueOf(maxRadius));
                    }

                    final int minRadius = 1;
                    if (radius < minRadius) {
                        radius = minRadius;
                        showMessage("command.levelledmobs.kill.near.invalid-radius-min",
                            "%minRadius%", String.valueOf(minRadius));
                    }

                    int killed = 0;
                    int skipped = 0;
                    final Collection<Entity> mobsToKill;
                    if (sender instanceof BlockCommandSender) {
                        final Block block = ((BlockCommandSender) sender).getBlock();
                        mobsToKill = block.getWorld()
                            .getNearbyEntities(block.getLocation(), radius, radius, radius);
                    } else {
                        mobsToKill = ((Player) sender).getNearbyEntities(radius, radius, radius);
                    }

                    for (final Entity entity : mobsToKill) {
                        if (!(entity instanceof LivingEntity)) continue;

                        final LivingEntity livingEntity = (LivingEntity) entity;
                        if (!main.levelInterface.isLevelled(livingEntity)) continue;

                        if (skipKillingEntity(main, livingEntity, rl)) {
                            skipped++;
                            continue;
                        }

                        livingEntity.setMetadata("noCommands",
                            new FixedMetadataValue(main, 1));

                        if (useNoDrops) {
                            livingEntity.remove();
                        } else {
                            livingEntity.setHealth(0.0);
                        }
                        killed++;
                    }

                    showMessage("command.levelledmobs.kill.near.success",
                        new String[]{"%killed%", "%skipped%", "%radius%"},
                        new String[]{String.valueOf(killed), String.valueOf(skipped),
                            String.valueOf(radius)}
                    );
                } else {
                    showMessage("command.levelledmobs.kill.near.usage");
                }

                break;
            default:
                showMessage("command.levelledmobs.kill.usage");
        }
    }

    @Nullable private RequestedLevel getLevelFromCommand(final @NotNull CommandSender sender,
        final String @NotNull [] args) {
        int rangeSpecifiedFlag = -1;

        for (int i = 0; i < args.length; i++) {
            if ("/levels".equalsIgnoreCase(args[i])) {
                rangeSpecifiedFlag = i + 1;
            }
        }

        if (rangeSpecifiedFlag <= 0) {
            return null;
        }

        final RequestedLevel rl = new RequestedLevel();
        if (args.length <= rangeSpecifiedFlag) {
            sender.sendMessage("No value was specified for /levels");
            rl.hadInvalidArguments = true;
            return rl;
        }

        final String value = args[rangeSpecifiedFlag];
        if (!rl.setLevelFromString(value)) {
            sender.sendMessage("Invalid number or range specified for /levels");
            rl.hadInvalidArguments = true;
        }

        return rl;
    }

    @Override
    public List<String> parseTabCompletions(final LevelledMobs main,
        final @NotNull CommandSender sender, final String[] args) {
        if (!sender.hasPermission("levelledmobs.command.kill")) {
            return Collections.emptyList();
        }

        boolean containsNoDrops = false;
        boolean containsLevels = false;

        for (final String arg : args) {
            if ("/nodrops".equalsIgnoreCase(arg)) {
                containsNoDrops = true;
            } else if ("/levels".equalsIgnoreCase(arg)) {
                containsLevels = true;
            }
        }

        if (args.length == 2) {
            return List.of("all", "near");
        }

        if (args[1].equalsIgnoreCase("all") && (args.length == 3 || args.length == 4)) {
            if (sender.hasPermission("levelledmobs.command.kill.all")) {
                final List<String> worlds = new LinkedList<>();

                if (!containsNoDrops) {
                    worlds.add("/nodrops");
                }
                if (!containsLevels) {
                    worlds.add("/levels");
                }
                if (args.length == 3) {
                    for (final World world : Bukkit.getWorlds()) {
                        worlds.add("*");
                        if (main.rulesManager.getRuleIsWorldAllowedInAnyRule(world)) {
                            worlds.add(world.getName());
                        }
                    }
                }

                return worlds;
            }
        }
        if (args[1].equalsIgnoreCase("near") && sender.hasPermission("levelledmobs.command.kill.near")) {
            if (args.length == 4 && "/levels".equalsIgnoreCase(args[3])) {
                return List.of("/levels");
            } else if (args.length == 3) {
                return Utils.oneToNine;
            }
        }

        final List<String> result = new LinkedList<>();
        if (!containsNoDrops) {
            result.add("/nodrops");
        }
        if (!containsLevels) {
            result.add("/levels");
        }

        return result;
    }

    private void parseKillAll(final @NotNull List<World> worlds, final LevelledMobs main,
        final boolean useNoDrops, final RequestedLevel rl) {
        int killed = 0;
        int skipped = 0;

        for (final World world : worlds) {
            for (final Entity entity : world.getEntities()) {
                if (!(entity instanceof LivingEntity)) {
                    continue;
                }
                final LivingEntity livingEntity = (LivingEntity) entity;
                if (!main.levelInterface.isLevelled(livingEntity)) {
                    continue;
                }

                if (skipKillingEntity(main, livingEntity, rl)) {
                    skipped++;
                    continue;
                }

                livingEntity.setMetadata("noCommands", new FixedMetadataValue(main, 1));

                if (useNoDrops) {
                    livingEntity.remove();
                } else {
                    livingEntity.setHealth(0.0);
                }

                killed++;
            }
        }

        showMessage("command.levelledmobs.kill.all.success",
            new String[]{"%killed%", "%skipped%", "%worlds%"},
            new String[]{String.valueOf(killed), String.valueOf(skipped),
                String.valueOf(worlds.size())}
        );
    }

    private boolean skipKillingEntity(final LevelledMobs main,
        final @NotNull LivingEntity livingEntity, final RequestedLevel rl) {
        if (livingEntity.getCustomName() != null && main.helperSettings.getBoolean(main.settingsCfg,
            "kill-skip-conditions.nametagged")) {
            return true;
        }

        if (rl != null) {
            final int mobLevel = main.levelInterface.getLevelOfMob(livingEntity);
            if (mobLevel < rl.getLevelMin() || mobLevel > rl.getLevelMax()) {
                return true;
            }
        }

        // Tamed
        if (livingEntity instanceof Tameable && ((Tameable) livingEntity).isTamed()
            && main.helperSettings.getBoolean(main.settingsCfg, "kill-skip-conditions.tamed")) {
            return true;
        }

        // Leashed
        if (livingEntity.isLeashed() && main.helperSettings.getBoolean(main.settingsCfg,
            "kill-skip-conditions.leashed")) {
            return true;
        }

        // Converting zombie villager
        return livingEntity.getType() == EntityType.ZOMBIE_VILLAGER &&
            ((ZombieVillager) livingEntity).isConverting() &&
            main.helperSettings.getBoolean(main.settingsCfg,
                "kill-skip-conditions.convertingZombieVillager");
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/commands/subcommands/KillSubcommand.kt
package io.github.arcaneplugins.levelledmobs.commands.subcommands

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.suggestion.Suggestions
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import com.mojang.brigadier.tree.LiteralCommandNode
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.MainCompanion
import io.github.arcaneplugins.levelledmobs.misc.RequestedLevel
import io.github.arcaneplugins.levelledmobs.util.MiscUtils
import io.github.arcaneplugins.levelledmobs.util.Utils
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import io.papermc.paper.command.brigadier.CommandSourceStack
import java.util.TreeSet
import java.util.concurrent.CompletableFuture
import org.bukkit.Bukkit
import org.bukkit.World
import org.bukkit.command.BlockCommandSender
import org.bukkit.command.CommandSender
import org.bukkit.entity.AbstractVillager
import org.bukkit.entity.Entity
import org.bukkit.entity.Hoglin
import org.bukkit.entity.Husk
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.PigZombie
import org.bukkit.entity.PiglinAbstract
import org.bukkit.entity.Player
import org.bukkit.entity.Skeleton
import org.bukkit.entity.Tameable
import org.bukkit.entity.Zombie
import org.bukkit.entity.ZombieVillager
import org.bukkit.metadata.FixedMetadataValue

/**
 * Allows you to kill LevelledMobs with various options including all levelled mobs, specific worlds
 * or levelled mobs in your proximity
 *
 * @author stumper66
 * @since 2.0
 */
object KillSubcommand : CommandBase("levelledmobs.command.kill") {
    override val description = "Various commands for killing LevelledMobs mobs."

    fun buildCommand() : LiteralCommandNode<CommandSourceStack> {
        return createLiteralCommand("kill")
            .executes { ctx -> processCmd(ctx, true)
                return@executes Command.SINGLE_SUCCESS
            }
            .then(createLiteralCommand("all")
                .then(createGreedyStringArgument("all")
                    .suggests { _, builder -> buildTabSuggestions(builder) }
                    .executes { ctx -> processCmd(ctx, true)
                        return@executes Command.SINGLE_SUCCESS })
            .executes { ctx -> processCmd(ctx, true)
                return@executes Command.SINGLE_SUCCESS
            })
            .then(createLiteralCommand("near")
                .then(createGreedyStringArgument("near")
                    .suggests { _, builder -> buildTabSuggestions(builder) }
                    .executes { ctx -> processCmd(ctx, false)
                        return@executes Command.SINGLE_SUCCESS })
            .executes { ctx -> processCmd(ctx, true)
                return@executes Command.SINGLE_SUCCESS
            })
            .build()
    }

    private fun processCmd(
        ctx: CommandContext<CommandSourceStack>,
        isAll: Boolean
    ){
        val sender = ctx.source.sender
        commandSender = sender
        //TODO: make this work in Folia
        if (LevelledMobs.instance.ver.isRunningFolia) {
            sender.sendMessage("Sorry this command doesn't work in Folia")
            return
        }

        val values = ctx.input
        if (values == null) {
            if (sender is Player) {
                if (isAll)
                    processKillAll(mutableListOf(sender.world), false, null)
                else
                    showMessage("command.levelledmobs.kill.near.usage")
            } else {
                showMessage(
                    if (isAll) "command.levelledmobs.kill.all.usage-console"
                    else "command.levelledmobs.kill.near.usage-console")
            }
            return
        }

        val args = splitStringWithQuotes(values, false)
        var useNoDrops = false
        val rl = getLevelFromCommand(sender, args)
        val newArgs = mutableListOf<String>()
        if (rl?.hadInvalidArguments ?: false) return

        var rlFlag = -1
        for (i in 3..<args.size) {
            val checkArg = args[i]
            if ("/nodrops".equals(checkArg, ignoreCase = true))
                useNoDrops = true
            else if ("/levels".equals(checkArg, ignoreCase = true))
                rlFlag = i + 1
            else if (i != rlFlag)
                newArgs.add(checkArg)
        }

        val options = Options(newArgs, useNoDrops, rl)

        if (isAll)
            processKillAll(sender, options)
        else
            processKillNear(sender, options)
    }

    private fun processKillAll(
        sender: CommandSender,
        opts: Options
    ){
        var useArg = ""
        for (i in 0..<opts.args.size){
            if (opts.args[i].startsWith("/")) continue
            useArg = opts.args[i]
        }

        if (useArg.isEmpty()) {
            if (sender is Player)
                processKillAll(mutableListOf(sender.world), opts.useNoDrops, opts.requestedLevel)
            else
                showMessage("command.levelledmobs.kill.all.usage-console")
        } else {
            if (useArg == "*") {
                processKillAll(Bukkit.getWorlds(), opts.useNoDrops, opts.requestedLevel)
                return
            }

            val world = Bukkit.getWorld(useArg)
            if (world == null) {
                showMessage(
                    "command.levelledmobs.kill.all.invalid-world", "%world%",
                    useArg
                )
                return
            }
            processKillAll(mutableListOf(world), opts.useNoDrops, opts.requestedLevel)
        }
    }

    private fun processKillNear(
        sender: CommandSender,
        opts: Options
    ){
        if (sender !is BlockCommandSender && sender !is Player) {
            showMessage("common.players-only")
            return
        }

        if (opts.args.isEmpty()){
            showMessage("command.levelledmobs.kill.near.usage")
            return
        }

        var radius: Int
        try {
            radius = opts.args[0].toInt()
        } catch (_: NumberFormatException) {
            showMessage("command.levelledmobs.kill.near.invalid-radius", "%radius%",
                opts.args.first()
            )
            return
        }

        val maxRadius = 1000
        if (radius > maxRadius) {
            radius = maxRadius
            showMessage("command.levelledmobs.kill.near.invalid-radius-max",
                "%maxRadius%", maxRadius.toString()
            )
        }

        val minRadius = 1
        if (radius < minRadius) {
            radius = minRadius
            showMessage("command.levelledmobs.kill.near.invalid-radius-min",
                "%minRadius%", minRadius.toString()
            )
        }

        var killed = 0
        var skipped = 0
        val mobsToKill: Collection<Entity>

        if (sender is BlockCommandSender) {
            val block = sender.block
            val finalRadius = MiscUtils.retrieveLoadedChunkRadius(block.location, radius.toDouble())
            mobsToKill = block.world
                .getNearbyEntities(block.location, finalRadius, finalRadius, finalRadius)
        } else {
            val finalRadius = MiscUtils.retrieveLoadedChunkRadius((sender as Player).location, radius.toDouble())
            mobsToKill = sender.getNearbyEntities(finalRadius, finalRadius, finalRadius)
        }

        for (entity in mobsToKill) {
            if (entity !is LivingEntity) continue

            if (!LevelledMobs.instance.levelInterface.isLevelled(entity)) continue

            if (skipKillingEntity(entity, opts.requestedLevel)) {
                skipped++
                continue
            }

            entity.setMetadata(
                "noCommands",
                FixedMetadataValue(LevelledMobs.instance, 1)
            )

            if (opts.useNoDrops)
                entity.remove()
            else
                entity.health = 0.0

            killed++
        }

        showMessage("command.levelledmobs.kill.near.success",
            mutableListOf("%killed%", "%skipped%", "%radius%"),
            mutableListOf(
                killed.toString(), skipped.toString(),
                radius.toString()
            )
        )
    }

    private fun buildTabSuggestions(
        builder: SuggestionsBuilder
    ): CompletableFuture<Suggestions> {

        val args = splitStringWithQuotes(builder.input, true)
        if (args.size == 2) return builder.buildFuture()
        val lastArg = args[args.size - 1]
        val prefix = StringBuilder()
        val existingItems = mutableListOf<String>()
        var containsNoDrops = false
        var containsLevels = false
        var containsWorld = false
        val hasRemaining = builder.remaining.isNotEmpty() && !builder.remaining.endsWith(' ')

        if ("/levels".equals(lastArg, ignoreCase = true)) {
            Utils.getOneToNineSuggestions(builder)
            return builder.buildFuture()
        }

        for (i in 3..<args.size) {
            val arg = args[i]

            existingItems.add(arg.lowercase())
            prefix.append(arg).append(' ')

            if ("/nodrops".equals(arg, ignoreCase = true))
                containsNoDrops = true
            else if ("/levels".equals(arg, ignoreCase = true))
                containsLevels = true
            else if (!arg.startsWith("/") && !hasRemaining && !Utils.isInteger(arg))
                containsWorld = true
        }

        if ("all".equals(args[2], ignoreCase = true) && args.size <= 7) {
            val commandsList: MutableSet<String> = TreeSet(String.CASE_INSENSITIVE_ORDER)
            commandsList.addAll(mutableListOf("/nodrops", "/levels"))
            if (args.isEmpty()) return builder.buildFuture()

            args.forEach { arg ->
                if (commandsList.contains(arg))
                    commandsList.remove(arg)
            }

            commandsList.forEach { command -> builder.suggest("$prefix$command") }

            if (!containsWorld) {
                for (world in Bukkit.getWorlds()) {
                    builder.suggest("*")
                    if (LevelledMobs.instance.rulesManager.getRuleIsWorldAllowedInAnyRule(world) &&
                        !existingItems.contains(world.name))
                            builder.suggest("$prefix${world.name}")
                }
            }

            return builder.buildFuture()
        }
        else if ("near".equals(args[2], ignoreCase = true)) {
            if (args.size == 3) {
                Utils.getOneToNineSuggestions(builder)
                return builder.buildFuture()
            }

            if ("/levels".equals(lastArg, ignoreCase = true)) {
                builder.suggest("/levels")
                return builder.buildFuture()
            }
        }

        if (!containsNoDrops)
            builder.suggest("/nodrops")

        if (!containsLevels)
            builder.suggest("/levels")

        return builder.buildFuture()
    }

    private fun getLevelFromCommand(
        sender: CommandSender,
        args: MutableList<String>
    ): RequestedLevel? {
        var rangeSpecifiedFlag = -1

        for (i in args.indices) {
            if ("/levels".equals(args[i], ignoreCase = true))
                rangeSpecifiedFlag = i + 1
        }

        if (rangeSpecifiedFlag <= 0)
            return null

        val rl = RequestedLevel()
        if (args.size <= rangeSpecifiedFlag) {
            sender.sendMessage("No value was specified for /levels")
            rl.hadInvalidArguments = true
            return rl
        }

        val value = args[rangeSpecifiedFlag]
        if (!rl.setLevelFromString(value)) {
            sender.sendMessage("Invalid number or range specified for /levels")
            rl.hadInvalidArguments = true
        }

        return rl
    }

    private fun processKillAll(
        worlds: MutableList<World>,
        useNoDrops: Boolean,
        rl: RequestedLevel?
    ) {
        LevelledMobs.instance.mobsQueueManager.clearQueue()
        var killed = 0
        var skipped = 0

        for (world in worlds) {
            for (entity in world.entities) {
                if (entity !is LivingEntity) continue

                if (!LevelledMobs.instance.levelInterface.isLevelled(entity))
                    continue

                if (skipKillingEntity(entity, rl)) {
                    skipped++
                    continue
                }

                entity.setMetadata("noCommands", FixedMetadataValue(LevelledMobs.instance, 1))

                if (useNoDrops)
                    entity.remove()
                else
                    entity.health = 0.0

                killed++
            }
        }

        showMessage("command.levelledmobs.kill.all.success",
            mutableListOf("%killed%", "%skipped%", "%worlds%"),
            mutableListOf(
                killed.toString(), skipped.toString(),
                worlds.size.toString()
            )
        )
    }

    private fun skipKillingEntity(
        livingEntity: LivingEntity,
        rl: RequestedLevel?
    ): Boolean {
        val skc = MainCompanion.instance.killSkipConditions
        @Suppress("DEPRECATION")
        if (livingEntity.customName != null && skc.isNametagged
        ) {
            return true
        }

        if (rl != null) {
            val mobLevel = LevelledMobs.instance.levelInterface.getLevelOfMob(livingEntity)
            if (mobLevel < rl.levelMin || mobLevel > rl.levelMax)
                return true
        }

        // Tamed
        if (livingEntity is Tameable && skc.isTamed)
            return true

        // Leashed
        if (livingEntity.isLeashed && skc.isLeashed)
            return true

        if (livingEntity is AbstractVillager && skc.isVillager)
            return true

        if (skc.entityTypes != null && !skc.entityTypes!!.isEmpty()){
            val et = skc.entityTypes!!
            if (et.includeAll) return false
            if (et.excludeAll) return true
            val lmEntity = LivingEntityWrapper.getInstance(livingEntity)
            try{
                if (et.isIncludedInList(lmEntity.nameIfBaby, lmEntity) ||
                    et.includedList.contains("baby_") && lmEntity.isBabyMob)
                    return true
            }
            finally {
                lmEntity.free()
            }
        }

        return (skc.isTransforming && isMobConverting(livingEntity))
    }

    private fun isMobConverting(mob: LivingEntity): Boolean{
        if (mob is Hoglin && mob.isConverting) return true
        if (mob is Husk && mob.isConverting) return true
        if (mob is PiglinAbstract && mob.isConverting) return true
        if (mob is PigZombie && mob.isConverting) return true
        if (mob is Skeleton && mob.isConverting) return true
        if (mob is Zombie && mob.isConverting) return true
        return (mob is ZombieVillager && mob.isConverting)
    }

    private class Options(
        val args: MutableList<String>,
        val useNoDrops: Boolean,
        val requestedLevel: RequestedLevel?
    )
}